### PR TITLE
OIST dataset loader

### PR DESCRIPTION
**Pull request by ProfFan**
WIP oist dataset loading.

### CODE REVIEW

**Comment by ProfFan** [comment]
Usage:

```
fan@3f7f64b97602:/workspaces/SwiftFusion$ swift run -c release OISTVisualizationTool view-frame --help
[2/2] Linking OISTVisualizationTool
USAGE: oist-visualization-tool view-frame [--dataset-location <dataset-location>] [--frame-id <frame-id>]

OPTIONS:
  --dataset-location <dataset-location>
                          Location of dataset folder which should contain `frames` and `frames_txt` (default: ./OIST_Data)
  --frame-id <frame-id>   Which frame to show (default: 0)
  -h, --help              Show help information.
```

**Comment by dellaert** [comment]
Why merge into ppca not master? Should ppca branch not be in a PR and merged, as well?

**Comment by dellaert** [comment]
Somehow all tests disappeared when I check this branch out

File: Examples/OISTVisualizationTool/main.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import ArgumentParser
```
Comment (dellaert) [inline Examples/OISTVisualizationTool/main.swift:1]: doc?

File: Sources/BeeDataset/OISTBeeVideo.swift [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  public var scale: Double = 2
+}
+
+fileprivate extension String.StringInterpolation {
+    mutating func appendInterpolation(_ val: Int, padding: UInt) {
+        appendLiteral(String(format: "%0\(padding)d", arguments: [val]))
+    }
+}
+
+fileprivate extension String {
```
Comment (dellaert) [inline Sources/BeeDataset/OISTBeeVideo.swift:98]: doc

File: Sources/BeeDataset/OISTBeeVideo.swift [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  /// The format of a track file is:
+  /// - Arbitrarily many lines "offset_x, offset_y, bee_type, x, y, angle".
+  public init?(directory: URL, deferLoadingFrames: Bool = true, fps: Int = 30) {
+    self.frames = []
+    self.labels = []
+    self.frameIds = []
+    self.directory = directory
+    self.fps = fps
+
+    let directoryContents = try! FileManager.default.contentsOfDirectory(at: directory.appendingPathComponent("frames"), includingPropertiesForKeys: nil)
```
Comment (dellaert) [inline Sources/BeeDataset/OISTBeeVideo.swift:130]: doc

File: Sources/BeeDataset/OISTBeeVideo.swift [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+      let name = f.deletingPathExtension().lastPathComponent
+      let ranges = name.ranges(of: #"([0-9]+)"#, options: .regularExpression)
+      if ranges.count == 2 {
+        return Int(name[ranges[1]])!
+      } else {
+        fatalError("Bad file name!")
+      }
+    }.sorted()
+
+    if !deferLoadingFrames {
```
Comment (dellaert) [inline Sources/BeeDataset/OISTBeeVideo.swift:141]: doc

File: Sources/BeeDataset/OISTBeeVideo.swift [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+      while let frame = loadFrame(self.frameIds[self.frames.count]) {
+        self.frames.append(frame)
+        if self.frameIds.count == self.frames.count {
+          break
+        }
+      }
+    }
+
+    func loadTrack(_ index: Int) -> [OISTBeeLabel]? {
```
Comment (dellaert) [inline Sources/BeeDataset/OISTBeeVideo.swift:150]: doc, rename to loadLabels(frame index:Int)

File: Sources/BeeDataset/OISTBeeVideo.swift [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    while let label = loadTrack(self.frameIds[self.labels.count]) {
+      self.labels.append(label)
+      if self.frameIds.count == self.labels.count {
+          break
+        }
+    }
+  }
+
+  /// Example: frame_30fps_003525.txt
+  public func loadFrame(_ index: Int) -> Tensor<Double>? {
```
Comment (dellaert) [inline Sources/BeeDataset/OISTBeeVideo.swift:183]: frame:

File: Sources/BeeDataset/OISTBeeVideo.swift [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    let path = self.directory!.appendingPathComponent("frames").appendingPathComponent("frame_\(fps)fps_\(index, padding: 6).png")
+    guard FileManager.default.fileExists(atPath: path.path) else { return nil }
+    let downsampler = AvgPool2D<Double>(poolSize: (scale, scale), strides: (scale, scale), padding: .same)
+    return downsampler(Tensor<Double>(ModelSupport.Image(contentsOf: path).tensor).expandingShape(at: 0)).squeezingShape(at: 0)
+  }
+  
+  private static func downloadDatasetIfNotPresent() -> URL {
```
Comment (dellaert) [inline Sources/BeeDataset/OISTBeeVideo.swift:190]: doc

File: Sources/BeeDataset/OISTBeeVideo.swift [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    }
+    while let label = loadTrack(self.frameIds[self.labels.count]) {
+      self.labels.append(label)
+      if self.frameIds.count == self.labels.count {
+          break
+        }
+    }
+  }
+
+  /// Example: frame_30fps_003525.txt
```
Comment (dellaert) [inline Sources/BeeDataset/OISTBeeVideo.swift:182]: doc

File: Sources/BeeDataset/Visualization.swift [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import SwiftFusion
```
Comment (dellaert) [inline Sources/BeeDataset/Visualization.swift:1]: doc, copyright

File: Sources/BeeDataset/Visualization.swift [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import TensorFlow
+import Plotly
+import ModelSupport
+import Foundation
+
+/// A Plotly figure that displays `frame`, with optional `boxes` overlaid.
+public func plotImagePlotly(
+  _ frame: Tensor<Double>, boxes: [(name: String, OrientedBoundingBox)] = [],
```
Comment (dellaert) [inline Sources/BeeDataset/Visualization.swift:9]: plot(image:, bboxes:)

**Review by dellaert — APPROVED** [review]
Cool !

**Pull request merged**